### DISCUSSION TITLE
Designing cameras for frigate with OpenIPC

### DISCUSSION THREAD

**Discussion by cat101** [blakeblackshear/frigate#7591] [Ask A Question]
Hi. I'm planning to flash several Smartware cameras with OpenIPC. Some of the features that I could set up are: multiple streams, motion detection, person detection (maybe), FTP upload and MQTT events (among many other options). I'm resource conscious since my setup is based on a Jetson Nano and I plan to have 10+ cameras.

Can you help me answer these questions:

1. I understand that low-res streams help reduce GPU usage for object detection. Still, two streams would increase WIFI channel utilization (vs one stream). If I want to preserve WIFI utilization, how bad would it be to resize the image on frigate? Is the CPU cost so bad compared to double streaming??  (I understand that with multiple streams only one gets decoded)
2. If I want to use Frigate as a dumb motion based DVR for some cameras, it seems I could use the new "[manual events API](https://github.com/blakeblackshear/frigate/pull/3184)" to upload events based on MQTT notifications from the camera.  Is there other way to integrate "sources of recording"? I also want to pull clips from an analog DVR
3. Each ffmpeg decoder instances uses 12% CPU on the Jetson. What if I wanted to only start decoding when the camera detects motion? Can I disable/enable decoding (even if it creates some detection lag)? I see on MQTT a motion and detection topic. Will them stop decoding.

Thanks for your help and ideas!

PS: @blakeblackshear Frigate is an amazing piece of software. It works beautifully!.  @madsciencetist thanks for the Jetpack4 build! I had already build ffmpeg with NVENC support and was starting to work on TRT when I saw your [PR](https://github.com/blakeblackshear/frigate/pull/6458). Thanks!

**Comment by madsciencetist** [comment]
You'll have your primary stream that you want to record. Let's say this is 5MP 30FPS. This stream is recorded without decoding, so the act of recording uses very little CPU.

Now, frigate needs to decode a stream for motion and object detection. By default it will decode the primary stream, downscale it to 1MP and subsample it to 5 FPS. Alternatively, you can use a secondary stream that is already low resolution and rate (which IP cameras call a "substream"). The CPU (and decoder) load is proportional to the resolution and framerate of the stream, so by using a 1MP 5FPS substream for detection instead of the 5MP 30FPS stream, decoding becomes 97% cheaper and downscaling becomes free. While there is the associated WiFi bandwidth increase, in this example it's only a ~3% increase!

That said, the tradeoff is entirely up to you. Both will work. I personally do not use a substream because my camera only supports substreams up to 0.3MP, which does not give my wide-angle camera enough pixels on target. If your CPU load is acceptable without substreams, then you don't need to use them - but with 10+ cameras on a Jetson Nano, I'd be looking to save all the compute resources that you can. You can test the compute load of 10 cameras without actually needing 10 cameras by configuring the same camera in frigate 10 times.

Note that the GPU usage for object detection is unrelated & unaffected. Entire image is decoded using the media engine -> downscaled to 1MP if necessary -> CPU motion detection detects regions of motion -> regions of motion are sent to GPU object detector. Replacing the first two steps with a low-res substream does not by itself change the size of the region that is passed to the object detector (but there are other parameters for that). Also note that you can't only start decoding when it detects motion because the motion detector operates on the decoded image. You could only start _recording_ when motion is detected, but recording doesn't require decoding. You can't start/stop decoding at runtime, and if that feature were implemented, you wouldn't be able to detect motion when not decoding.

I don't quite understand question (2), but you can definitely "use Frigate as a dumb motion-based DVR for some cameras". Simply configure those cameras to retain recordings when motion is detected, and don't configure them for object detection. Motion detection will trigger MQTT notifications that you can act on downstream. I don't know what to tell you about pulling clips from an analog DVR though.

**Comment by cat101** [comment]
Thanks @madsciencetist for your prompt reply. Using what you shared with me, let me rephrase the part were I was not clear: Again, the goal is to save resources on the Jetson by offloading work to the cameras that run my custom firmware. What if I record & do motion detection on the camera. Could I write a batch process to import those motion events from the camera into Frigate every 10 min? (the DVR question boils down to the same idea of "side loading" events and recordings to Frigate)

BYW: Thanks for the substream explanation, I had not realized that I can create a custom  5FPS stream for frigate. If I do 580p this would only need 100Kbps on my wifi and as you mention saved a ton of CPU spent on decoding (which seems to be my bottleneck for 10 cameras)

**Reply by madsciencetist** [reply]
Yes, you could write a script to import those motion events into Frigate, but at that point, you're writing the code, so you can do anything you want. There isn't any sort of recordings importer built in. You'll have to get familiar with Frigate's event database and recordings manager.
